lib/templates.js

```diff
@@ -184,15 +184,59 @@ const createGroupsTemplateForService = (serviceName, serviceDefinition, appConfi
 
 /**
  * This is a template function to create EntityType object for EntityTypes Array.
+ * Ensures correct visibility assignment based on referenced resources.
+ *
  * @param { object } appConfig  The configuration object.
- * @param { Array } packageIds The identifiers of packages.
- * @param { object } entity
+ * @param { Array } packageIds The available package identifiers.
+ * @param { object } entity The entity definition.
  * @returns { object } An object for the EntityType.
  */
 const createEntityTypeTemplate = (appConfig, packageIds, entity) => {
     const ordExtensions = readORDExtensions(entity);
+    const visibility = ordExtensions.visibility || RESOURCE_VISIBILITY.public;
+
+    // Check whether the EntityType is referenced by a restrictive API
+    const referencedByRestrictedAPI = (Array.isArray(appConfig.apiResources) ? appConfig.apiResources : []).some(api =>
+        api.entityTypeMappings?.some(mapping =>
+            mapping.entityTypeTargets.some(target => target.ordId === entity.ordId)
+        ) && [RESOURCE_VISIBILITY.private, RESOURCE_VISIBILITY.internal].includes(api.visibility)
+    );
+
+   // Check whether the EntityType is referenced by a restrictive event resource
+    const referencedByRestrictedEvent = (Array.isArray(appConfig.eventResources) ? appConfig.eventResources : []).some(event =>
+        event.entityTypeMappings?.some(mapping =>
+            mapping.entityTypeTargets.some(target => target.ordId === entity.ordId)
+        ) && [RESOURCE_VISIBILITY.private, RESOURCE_VISIBILITY.internal].includes(event.visibility)
+    );
+
+    // Check whether the EntityType is referenced by a restrictive DataProduct
+    const referencedByRestrictedDataProduct = (Array.isArray(appConfig.dataProducts) ? appConfig.dataProducts : []).some(dp =>
+        dp.entityTypes?.some(target => target === entity.ordId)
+        && [RESOURCE_VISIBILITY.private, RESOURCE_VISIBILITY.internal].includes(dp.visibility)
+    );
+
+    // Adopt the most restrictive visibility
+    let finalVisibility = visibility;
+    if (referencedByRestrictedAPI || referencedByRestrictedEvent || referencedByRestrictedDataProduct) {
+        if (referencedByRestrictedAPI && appConfig.apiResources.some(api => api.visibility === RESOURCE_VISIBILITY.private) ||
+            referencedByRestrictedEvent && appConfig.eventResources.some(event => event.visibility === RESOURCE_VISIBILITY.private) ||
+            referencedByRestrictedDataProduct && appConfig.dataProducts.some(dp => dp.visibility === RESOURCE_VISIBILITY.private)) {
+            finalVisibility = RESOURCE_VISIBILITY.private;
+        } else {
+            finalVisibility = RESOURCE_VISIBILITY.internal;
+        }
```

Comment:
The visibility logic should be the other way around. If it is referenced by any public resource, the entity is public. if it is referenced by no public, but any internal, it is internal, if it is reference by only private resources, it is private.

Reply:
and this logic can be combined with the functions above

Reply:
and I miss clear test cases that check this logic. they might exist, but I don't see them directly.